File: days/day9.py
#!/usr/bin/env python


def part1(data):
    xs = list(map(int, data.strip()))
    disk = list()
    i = 0
    for x,s in zip(xs[::2], xs[1::2]):
        disk.extend([i] * x)
        disk.extend([None] * s)
        i += 1
    if len(xs) % 2:
        disk.extend([i] * xs[-1])
    r = len(disk) - 1
    l = disk.index(None)
    while l < r:
        x = disk[r]
        if x is not None:
            disk[l] = x
            disk[r] = None
            l = disk.index(None, l+1)
        r -= 1
    ans = sum(i*(x or 0) for i,x in enumerate(disk))
    return ans


def part2(data):
    xs = list(map(int, data.strip()))
    frag = xs[::2]
    free = xs[1::2]
    disk = [list() for _ in range(len(xs))]

    for i in range(len(frag))[::-1]:
        n = frag[i]
        for j in range(i):
            if n <= free[j]:
                free[j] -= n
                free[i-1] += n
                disk[1+j*2].append(i)
                break
        else:
            disk[i*2].append(i)

    ans = 0
    i = 0
    for j,ps in enumerate(disk):
        for x in ps:
            for _ in range(frag[x]):
                ans += i * x
                i += 1
        if j % 2:
            i += free[j//2]
    return ans


data = '''
2333133121414131402
'''
assert part1(data) == 1928
assert part2(data) == 2858


data = open('day9.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)

File: days/day9.jl
#!/usr/bin/env julia


const Block = Tuple{Int, Int}
const FileBlock = Tuple{Int, Int, Int}


function parsedisk(text)
    files = Vector{FileBlock}()
    space = Vector{Block}()
    state = 0
    fileid = 0
    i = 0
    for n in parse.(Int, collect(chomp(data)))
        if state == 0
            n > 0 && push!(files, (i, n, fileid))
            fileid += 1
        else
            n > 0 && push!(space, (i, n))
        end
        i += n
        state = 1 - state
    end
    return (files, space)
end


function _defrag!(disk)
    files, space = disk
    for k in reverse(1:length(files))
        (j, n, x) = files[k]
        for (q, (i, m)) in enumerate(space)
            j <= i && break
            if m >= n
                files[k] = (i, n, x)
                space[q] = (i + n, m - n)
                break
            end
        end
    end
end


function _checksum(disk)
    files, _ = disk
    return sum(x * (n * i + (n - 1) * n ÷ 2)
        for (i, n, x) in files)
end


function _flatten(disk)
    files, space = disk
    space = [(i+j-1, 1) for (i,n) in space for j in 1:n]
    files = [(i+j-1, 1, x) for (i,n,x) in files for j in 1:n]
    return (files, space)
end


function part1(data)
    disk = _flatten(parsedisk(data))
    _defrag!(disk)
    ans = _checksum(disk)
    return ans
end


function part2(data)
    disk = parsedisk(data)
    _defrag!(disk)
    ans = _checksum(disk)
    return ans
end


data = """
2333133121414131402
"""
@assert part1(data) == 1928
@assert part2(data) == 2858


data = readchomp("day9.in")
println("part1: ", part1(data))
println("part2: ", part2(data))
